signbank/dictionary/csv_import.py

```diff
@@ -0,0 +1,705 @@
+                    if not any(row[0] in s for s in glosses_new):
+                        glosses_new.append(tuple(row))
+                except IndexError:
+                    # If row[0] does not exist, continue to next iteration of loop.
+                    continue
+
+            # Store dataset's id and the list of glosses to be added in session.
+            request.session['dataset_id'] = dataset.id
+            request.session['glosses_new'] = glosses_new
+
+            return render(request, 'dictionary/import_gloss_csv_confirmation.html',
+                          {'glosses_new': glosses_new,
+                           'glosses_exists': glosses_exists,
+                           'dataset': dataset, })
+        else:
+            # If form is not valid, set a error message and return to the original form.
+            messages.add_message(request, messages.ERROR, _('The provided CSV-file does not meet the requirements '
+                                                            'or there is some other problem.'))
+            return render(request, 'dictionary/import_gloss_csv.html', {'import_csv_form': form}, )
+    else:
+        # If request type is not POST, return to the original form.
+        csv_form = CSVUploadForm()
+        allowed_datasets = get_objects_for_user(request.user, 'dictionary.view_dataset')
+        # Make sure we only list datasets the user has permissions to.
+        csv_form.fields["dataset"].queryset = csv_form.fields["dataset"].queryset.filter(
+            id__in=[x.id for x in allowed_datasets])
+        return render(request, "dictionary/import_gloss_csv.html",
+                      {'import_csv_form': csv_form}, )
+
+
+@login_required
+@permission_required('dictionary.import_csv')
+def confirm_import_gloss_csv(request):
+    """This view adds the data to database if the user confirms the action"""
+    if request.method == 'POST':
+        if 'cancel' in request.POST:
+            # If user cancels adding data, flush session variables
+            if 'dataset_id' in request.session: del request.session['dataset_id']
+            if 'glosses_new' in request.session: del request.session['glosses_new']
+            # Set a message to be shown so that the user knows what is going on.
+            messages.add_message(request, messages.WARNING, _('Cancelled adding CSV data.'))
+            return HttpResponseRedirect(reverse('dictionary:import_gloss_csv'))
+
+        elif 'confirm' in request.POST:
+            glosses_added = []
+            dataset = None
+            if 'glosses_new' and 'dataset_id' in request.session:
+                dataset = Dataset.objects.get(id=request.session['dataset_id'])
+                for gloss in request.session['glosses_new']:
+
+                    # If the Gloss does not already exist, continue adding.
+                    if not Gloss.objects.filter(dataset=dataset, idgloss=gloss[0]).exists():
```

Comment:
**suggestion**: gather the full set of `idgloss` values for this import first, then filter for the ids that are already in-use

```python
in_use = Gloss.objects.filter(
    dataset=dataset,
    idgloss__in=[g[0] for g in request.session['glosses_new']]
).values_list('idgloss', flat=True)
```